### PR TITLE
WIP using a loading store and having the apihandler to communicate wi…

### PR DESCRIPTION
**Pull request by SoLetsDev**
…th the store to handle loading states

---
name: Pull Request Template
about: Template for creating pull requests
---

## Title
ECER-1704: 
- Using a loading store to track whether we are waiting for a submission before proceeding with submit. 
- Removed mandatory error that was being thrown in the resultsHandler. 

## Checklist
- [x] I have tested these changes locally.
- [ ] Changes to backend endpoints are covered by passing integration tests.
- [ ] I have added or updated the necessary documentation.

## Screenshots (if applicable)
image

### CODE REVIEW

File: src/ECER.Clients.RegistryPortal/ecer.clients.registryportal.client/src/store/loading.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import type { OperationMethods } from "@/types/openapi";
+
+export type LoadingOperation = keyof OperationMethods;
+
+export type LoadingState = {
+  [key in LoadingOperation]: boolean;
+};
+
+export const useLoadingStore = defineStore("loading", {
+  state: (): LoadingState => ({
```
Comment (pcatkins) [inline src/ECER.Clients.RegistryPortal/ecer.clients.registryportal.client/src/store/loading.ts:16]: Nice use of `OperationMethods` — wonder if we can get away from setting initial object of flags? I.e. LoadingState would be typed as you have it but initialized to `{}`, on get/set check `'{operation_method: OperationMethods}' in this` or something like that... Then this should just work as we add/update/remove endpoints.

**Review by pcatkins — CHANGES_REQUESTED** [review]
Looks awesome, nice use of `OperationMethods`. Small improvement mentioned in comments.

File: src/ECER.Clients.RegistryPortal/ecer.clients.registryportal.client/src/store/loading.ts (hunk 1 above)
Comment (SoLetsDev) [reply src/ECER.Clients.RegistryPortal/ecer.clients.registryportal.client/src/store/loading.ts:16]: Yeah I think we can, let me play around. I think I need to do something with checking if a property exists then create it if it doesn't.

File: src/ECER.Clients.RegistryPortal/ecer.clients.registryportal.client/src/store/loading.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  }),
+
+  getters: {
+    isLoading() {
+      return (key: LoadingOperation) => this.loadingStates[key];
+    },
+  },
+  actions: {
+    async setLoading(key: LoadingOperation, loading: boolean) {
+      this.$patch({ loadingStates: { [key]: loading } });
```
Comment (SoLetsDev) [inline src/ECER.Clients.RegistryPortal/ecer.clients.registryportal.client/src/store/loading.ts:31]: patch is interesting, it allows us to just set the key and it'll retain all the other keys that were in there previously. 

it's similar to what we did in wizard 
this.$state.wizardData = {...this.$state.wizardData, newInfo: info};

File: src/ECER.Clients.RegistryPortal/ecer.clients.registryportal.client/src/store/loading.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+export const useLoadingStore = defineStore("loading", {
+  state: (): LoadingStateConfig => ({
+    loadingStates: {} as LoadingState,
+  }),
+
+  getters: {
+    isLoading() {
+      return (key: LoadingOperation) => this.loadingStates[key];
```
Comment (pcatkins) [inline src/ECER.Clients.RegistryPortal/ecer.clients.registryportal.client/src/store/loading.ts:26]: Do we need to check if loadingStates has the key? I think you can set the isLoading function to return type which may then complain that `this.loadingStates[key]` can return `undefined`

i.e.
```
  getters: {
    isLoading(): boolean {
      return (key: LoadingOperation) => this.loadingStates[key];
      ...
```

Comment (SoLetsDev) [reply src/ECER.Clients.RegistryPortal/ecer.clients.registryportal.client/src/store/loading.ts]: yeah it maintains all the other keys. I tested this by adding another property when we hit submit "draftapplication_put"

and I see that the key stays there 

image

File: src/ECER.Clients.RegistryPortal/ecer.clients.registryportal.client/src/store/loading.ts (hunk 2 above)
Comment (SoLetsDev) [reply src/ECER.Clients.RegistryPortal/ecer.clients.registryportal.client/src/store/loading.ts:31]: yeah it maintains all the other keys. I tested this by adding another property when we hit submit "draftapplication_put"

and I see that the key stays there 

image

**Pull request merged**